src/core/vanilla.ts

```diff
@@ -18,45 +18,82 @@ const hasInitialValue = <T extends Atom<unknown>>(
 type Revision = number
 type ReadDependencies = Map<AnyAtom, Revision>
 
+// immutable atom state
 export type AtomState<Value = unknown> = {
-  re?: Error // read error
-  rp?: Promise<void> // read promise
-  wp?: Promise<void> // write promise
+  e?: Error // read error
+  p?: Promise<void> // read promise
+  w?: Promise<void> // write promise
   v?: Value
   r: Revision
   d: ReadDependencies
 }
 
 type AtomStateMap = WeakMap<AnyAtom, AtomState>
 
-type UseAtomSymbol = symbol
-type Dependents = Set<AnyAtom | UseAtomSymbol>
-type MountedMap = Map<AnyAtom, [Dependents, OnUnmount | void]>
+type Listeners = Set<() => void>
+type Dependents = Set<AnyAtom>
+type Mounted = {
+  l: Listeners
+  d: Dependents
+  u: OnUnmount | void
+}
+
+type MountedMap = Map<AnyAtom, Mounted>
 
+type StateVersion = number
+
+// immutable map
```

Comment:
For future contributors, I think the concept of `WorkInProgress` should be explained in a comment.

Reply:
Totally. Let me add something. (I wanted to remove the concept fully, which didn't work in my first trial.)

Reply:
de177df